File: src/htomd/_selection.py
"""Select content using cached evidence, then clean only the selected region."""

from __future__ import annotations

import math
import re
from dataclasses import dataclass

from ._models import Diagnostics
from ._tree import Node, elements, postorder, walk

INERT = frozenset(
    [
        "head",
        "title",
        "meta",
        "link",
        "base",
        "script",
        "style",
        "template",
        "noscript",
        "iframe",
        "object",
        "embed",
        "svg",
        "canvas",
        "nav",
        "button",
        "input",
        "select",
        "textarea",
        "dialog",
    ]
)
NAV_ROLES = frozenset({"navigation", "banner", "contentinfo", "menu", "menubar", "dialog"})
STRUCTURES = frozenset({"ul", "ol", "dl", "table", "blockquote", "pre"})
CONTAINERS = frozenset({"article", "main", "section", "div", "body", "#document"})
EVIDENCE = frozenset({"p", "pre", "li", "dt", "dd", "td", "th", "blockquote"})
HEADINGS = frozenset({"h1", "h2", "h3", "h4", "h5", "h6"})
NEGATIVE = frozenset(
    [
        "advertisement",
        "ads",
        "advert",
        "promo",
        "promotion",
        "related",
        "share",
        "sharing",
        "social",
        "cookie",
        "consent",
        "newsletter",
        "comments",
        "comment",
        "sidebar",
        "breadcrumb",
        "breadcrumbs",
        "pagination",
        "toolbar",
        "footer",
        "banner",
        "dropdown",
        "catlinks",
        "menu",
        "pager",
        "teaser",
        "toc",
        "well",
    ]
)
POSITIVE = frozenset(
    {"article", "content", "main", "post", "entry", "story", "text", "documentation"}
)
REFERENCES = frozenset({"footnotes", "references", "endnotes", "bibliography"})
PUNCTUATION = re.compile(r"[,.;:!?。，；：！？،؛]")
HIDDEN_STYLE = re.compile(
    r"(?:^|;)\s*(?:display\s*:\s*none|visibility\s*:\s*(?:hidden|collapse))\s*(?:!important\s*)?(?:;|$)",
    re.I,
)
# Sibling recovery needs modest support, and can never escape its parent container.
SIBLING_SCORE_RATIO = 0.18
MAX_FALLBACK_BLOCKS = 256


@dataclass(slots=True)
class Stats:
    characters: int = 0
    linked: int = 0
    punctuation: int = 0
    blocks: int = 0
    headings: int = 0
    code: int = 0
    cells: int = 0
    images: int = 0
    controls: int = 0

    @property
    def density(self) -> float:
        return self.linked / max(1, self.characters)


def hints(node: Node) -> set[str]:
    value = node.attrs.get("class", "") + " " + node.attrs.get("id", "")
    value = re.sub(r"([a-z])([A-Z])", r"\1 \2", value)
    return set(re.findall(r"[a-z0-9]+", value.lower()))


def explicitly_hidden(node: Node) -> bool:
    return (
        "hidden" in node.attrs
        or node.attrs.get("aria-hidden", "").lower() == "true"
        or bool(HIDDEN_STYLE.search(node.attrs.get("style", "")))
    )


def visible_tree(root: Node) -> Node:
    """Copy visible content so metadata remains available on the original tree."""
    result = Node(root.tag, root.attrs.copy())
    stack = [(root, result, False)]
    while stack:
        original, target, local = stack.pop()
        local = local or original.tag in {"article", "main"} or original.attrs.get("role") == "main"
        for child in original.children:
            if isinstance(child, str):
                target.children.append(child)
                continue
            if excluded(child, local):
                continue
            copied = Node(child.tag, child.attrs.copy(), parent=target)
            target.children.append(copied)
            stack.append((child, copied, local))
    return result


def excluded(node: Node, local: bool) -> bool:
    if node.tag in INERT or explicitly_hidden(node):
        return True
    if node.tag == "a" and "headerlink" in hints(node):
        return node.attrs.get("href", "").startswith("#")
    role = node.attrs.get("role", "").lower()
    if role in NAV_ROLES:
        return True
    return node.tag in {"header", "footer"} and not local


def statistics(root: Node) -> dict[Node, Stats]:
    result: dict[Node, Stats] = {}
    for node in postorder(root):
        stat = Stats()
        for child in node.children:
            if isinstance(child, str):
                stat.characters += len(child.strip())
                stat.punctuation += len(PUNCTUATION.findall(child))
            else:
                other = result[child]
                for field in Stats.__slots__:
                    setattr(stat, field, getattr(stat, field) + getattr(other, field))
        stat.blocks += int(node.tag in EVIDENCE and stat.characters > 0)
        stat.headings += int(node.tag in HEADINGS and stat.characters > 0)
        stat.code += int(node.tag == "pre")
        stat.cells += int(node.tag in {"td", "th"})
        stat.images += int(node.tag == "img" and bool(node.attrs.get("src")))
        stat.controls += int(node.tag in {"form", "button", "input", "select", "textarea"})
        if node.tag == "a":
            stat.linked = stat.characters
        result[node] = stat
    return result


def conditional_clutter(node: Node, stat: Stats) -> bool:
    if node.tag in {"#document", "html", "body", "main", "article"}:
        return False
    tokens = hints(node)
    if tokens & REFERENCES or node.attrs.get("role") == "note":
        return False
    if not tokens & NEGATIVE or node.tag in EVIDENCE | HEADINGS:
        return False
    # Class hints alone never delete a node: require structural corroboration.
    if stat.density > 0.35 or stat.controls:
        return True
    if "footer" in tokens and stat.density > 0.15:
        return True
    if node.tag == "aside" and not stat.code and not stat.cells:
        return True
    if stat.blocks == 0 and stat.characters < 180:
        return True
    if tokens & {"comments", "comment"}:
        children = elements(node)
        return sum(bool(hints(child) & {"comment", "reply"}) for child in children) >= 2
    return False


def clean(root: Node) -> tuple[Node, int]:
    stats = statistics(root)
    removed = 0
    for node in walk(root):
        kept: list[Node | str] = []
        for child in node.children:
            if isinstance(child, Node) and conditional_clutter(child, stats[child]):
                removed += 1
            else:
                kept.append(child)
        node.children = kept
    return root, removed


def plausible(node: Node, stat: Stats, *, semantic: bool = False) -> bool:
    if not stat.characters:
        return bool(semantic and stat.images)
    if stat.code or stat.cells:
        return True
    if node.tag == "p" and stat.characters > stat.linked:
        return True
    if stat.density >= 0.8:
        return bool(semantic and stat.headings and stat.blocks)
    return node.tag in CONTAINERS | EVIDENCE | HEADINGS | STRUCTURES or semantic


def scores(root: Node, stats: dict[Node, Stats], *, relaxed: bool) -> dict[Node, float]:
    result: dict[Node, float] = {}
    for node in walk(root):
        stat = stats[node]
        if node.tag not in CONTAINERS or not plausible(node, stat):
            continue
        tokens = hints(node)
        score = math.sqrt(stat.characters) + min(stat.blocks, 30) * 2
        score += min(stat.punctuation, 40) * 0.25 + min(stat.code + stat.cells, 12) * 2
        score += 8 * bool(tokens & POSITIVE)
        if not relaxed:
            score -= 18 * bool(tokens & NEGATIVE)
        result[node] = score * (1 - stat.density) ** 2
    # Leaf evidence flows to nearby ancestors, without repeatedly counting nested blocks.
    for node in walk(root):
        if node.tag not in EVIDENCE or any(child.tag in EVIDENCE for child in elements(node)):
            continue
        stat = stats[node]
        support = (1 + min(stat.characters / 100, 3)) * (1 - stat.density)
        ancestor = node.parent
        for weight in (1.0, 0.5, 0.25):
            if ancestor is None:
                break
            if ancestor in result:
                result[ancestor] += support * weight
            ancestor = ancestor.parent
    return result


def semantic_candidate(
    root: Node, stats: dict[Node, Stats], ranked: dict[Node, float]
) -> Node | None:
    candidates = []
    landmarks = []
    for node in walk(root):
        landmark = node.tag == "main" or node.attrs.get("role") == "main"
        if not (landmark or node.tag == "article"):
            continue
        if not plausible(node, stats[node], semantic=True):
            continue
        if conditional_clutter(node, stats[node]):
            continue
        if "teaser" in hints(node) and stats[node].density > 0.1:
            continue
        candidates.append(node)
        if landmark:
            landmarks.append(node)
    candidates = landmarks or candidates
    if not candidates:
        return None
    return max(candidates, key=lambda node: ranked.get(node, 0))


def recover_siblings(
    winner: Node, stats: dict[Node, Stats], ranked: dict[Node, float]
) -> list[Node]:
    parent = winner.parent
    if parent is None:
        return [winner]
    threshold = max(5, ranked.get(winner, 0) * SIBLING_SCORE_RATIO)
    selected = []
    for sibling in elements(parent):
        stat = stats[sibling]
        if conditional_clutter(sibling, stat):
            continue
        supported = ranked.get(sibling, 0) >= threshold and stat.density < 0.5
        heading = sibling.tag in HEADINGS and stat.density < 0.5
        paragraph = sibling.tag == "p" and stat.density < 0.25 and bool(stat.characters)
        if sibling is winner or supported or heading or paragraph:
            selected.append(sibling)
    return selected


def fallback(root: Node, stats: dict[Node, Stats]) -> list[Node]:
    selected: list[Node] = []
    stack = [root]
    while stack and len(selected) < MAX_FALLBACK_BLOCKS:
        node = stack.pop()
        stat = stats[node]
        if conditional_clutter(node, stat):
            continue
        if (
            node.tag in EVIDENCE | HEADINGS | STRUCTURES
            and plausible(node, stat)
            or not elements(node)
            and stat.characters
            and stat.density < 0.5
        ):
            selected.append(node)
        else:
            stack.extend(reversed(elements(node)))
    return selected


def select(root: Node) -> tuple[list[Node], Diagnostics]:
    visible, removed = clean(visible_tree(root))
    stats = statistics(visible)
    ranked = scores(visible, stats, relaxed=False)
    notes = [f"Removed {removed} conditionally identified clutter blocks."] if removed else []
    winner = semantic_candidate(visible, stats, ranked)
    if winner is not None:
        return [winner], Diagnostics("semantic", tuple(notes))
    candidates = {
        node: score for node, score in ranked.items() if node.tag not in {"body", "#document"}
    }
    if not candidates or max(candidates.values()) < 5:
        notes.append("Retried selection once with relaxed class penalties.")
        ranked = scores(visible, stats, relaxed=True)
        candidates = {
            node: score for node, score in ranked.items() if node.tag not in {"body", "#document"}
        }
    if candidates and max(candidates.values()) >= 5:
        winner = max(candidates, key=lambda node: candidates[node])
        return recover_siblings(winner, stats, ranked), Diagnostics("scored", tuple(notes))
    recovered = fallback(visible, stats)
    notes.append(
        "Recovered plausible blocks." if recovered else "No relevant visible content found."
    )
    return recovered, Diagnostics("fallback" if recovered else "none", tuple(notes))

File: rust/src/selection.rs
use crate::{
    Diagnostics, Strategy,
    tree::{Id, Node, Tree, has},
};
use std::collections::{HashMap, HashSet};
const INERT: &str = "head title meta link base script style template noscript iframe object embed svg canvas nav button input select textarea dialog";
const NEGATIVE: &str = "advertisement ads advert promo promotion related share sharing social cookie consent newsletter comments comment sidebar breadcrumb breadcrumbs pagination toolbar footer banner dropdown catlinks menu pager teaser toc well";
const POSITIVE: &str = "article content main post entry story text documentation";
const REFERENCES: &str = "footnotes references endnotes bibliography";
const CONTAINERS: &str = "article main section div body #document";
const EVIDENCE: &str = "p pre li dt dd td th blockquote";
const STRUCTURES: &str = "ul ol dl table blockquote pre";
pub const HEADINGS: &str = "h1 h2 h3 h4 h5 h6";
fn hints(n: &Node) -> HashSet<String> {
    let value = format!("{} {}", n.attr("class"), n.attr("id"));
    let mut expanded = String::new();
    let mut previous = ' ';
    for c in value.chars() {
        if previous.is_ascii_lowercase() && c.is_ascii_uppercase() {
            expanded.push(' ');
        }
        expanded.push(c);
        previous = c;
    }
    expanded
        .to_lowercase()
        .split(|c: char| !c.is_ascii_lowercase() && !c.is_ascii_digit())
        .filter(|s| !s.is_empty())
        .map(str::to_owned)
        .collect()
}
fn intersects(tokens: &HashSet<String>, words: &str) -> bool {
    words.split_ascii_whitespace().any(|w| tokens.contains(w))
}
#[derive(Clone, Copy, Default)]
struct Stats {
    characters: usize,
    linked: usize,
    punctuation: usize,
    blocks: usize,
    headings: usize,
    code: usize,
    cells: usize,
    images: usize,
    controls: usize,
}
impl Stats {
    fn density(self) -> f64 {
        self.linked as f64 / self.characters.max(1) as f64
    }
}
fn hidden_style(style: &str) -> bool {
    style.split(';').any(|declaration| {
        let Some((key, value)) = declaration.split_once(':') else {
            return false;
        };
        let key = key.trim().to_ascii_lowercase();
        let lower = value.to_ascii_lowercase();
        let value = lower
            .trim()
            .strip_suffix("!important")
            .unwrap_or(lower.trim())
            .trim();
        key == "display" && value == "none"
            || key == "visibility" && matches!(value, "hidden" | "collapse")
    })
}
fn excluded(n: &Node, local: bool) -> bool {
    if has(INERT, &n.tag)
        || n.attrs.contains_key("hidden")
        || n.attr("aria-hidden").eq_ignore_ascii_case("true")
        || hidden_style(n.attr("style"))
    {
        return true;
    }
    if n.tag == "a" && hints(n).contains("headerlink") {
        return n.attr("href").starts_with('#');
    }
    if has(
        "navigation banner contentinfo menu menubar dialog",
        &n.attr("role").to_lowercase(),
    ) {
        return true;
    }
    has("header footer", &n.tag) && !local
}
fn visible(tree: &mut Tree) {
    let mut stack = vec![(0, false)];
    while let Some((id, local)) = stack.pop() {
        let n = &tree.nodes[id];
        let local = local || has("article main", &n.tag) || n.attr("role") == "main";
        let mut kept = Vec::new();
        for &c in &n.children {
            if !tree.nodes[c].tag.is_empty() {
                if excluded(&tree.nodes[c], local) {
                    continue;
                }
                stack.push((c, local));
            }
            kept.push(c);
        }
        tree.nodes[id].children = kept;
    }
}
fn statistics(tree: &Tree) -> Vec<Stats> {
    let mut result = vec![Stats::default(); tree.nodes.len()];
    for id in tree.walk(0).into_iter().rev() {
        let n = &tree.nodes[id];
        let mut s = Stats::default();
        for &c in &n.children {
            let child = &tree.nodes[c];
            if child.tag.is_empty() {
                s.characters += child.text.trim().chars().count();
                s.punctuation += child
                    .text
                    .chars()
                    .filter(|&c| ",.;:!?。，；：！？،؛".contains(c))
                    .count();
            } else {
                let o = result[c];
                s.characters += o.characters;
                s.linked += o.linked;
                s.punctuation += o.punctuation;
                s.blocks += o.blocks;
                s.headings += o.headings;
                s.code += o.code;
                s.cells += o.cells;
                s.images += o.images;
                s.controls += o.controls;
            }
        }
        s.blocks += usize::from(has(EVIDENCE, &n.tag) && s.characters > 0);
        s.headings += usize::from(has(HEADINGS, &n.tag) && s.characters > 0);
        s.code += usize::from(n.tag == "pre");
        s.cells += usize::from(has("td th", &n.tag));
        s.images += usize::from(n.tag == "img" && !n.attr("src").is_empty());
        s.controls += usize::from(has("form button input select textarea", &n.tag));
        if n.tag == "a" {
            s.linked = s.characters;
        }
        result[id] = s;
    }
    result
}
fn clutter(tree: &Tree, id: Id, s: Stats) -> bool {
    let n = &tree.nodes[id];
    if has("#document html body main article", &n.tag) {
        return false;
    }
    let tokens = hints(n);
    if intersects(&tokens, REFERENCES) || n.attr("role") == "note" {
        return false;
    }
    if !intersects(&tokens, NEGATIVE) || has(EVIDENCE, &n.tag) || has(HEADINGS, &n.tag) {
        return false;
    }
    if s.density() > 0.35
        || s.controls > 0
        || tokens.contains("footer") && s.density() > 0.15
        || n.tag == "aside" && s.code == 0 && s.cells == 0
        || s.blocks == 0 && s.characters < 180
    {
        return true;
    }
    if intersects(&tokens, "comments comment") {
        return tree
            .elements(id)
            .iter()
            .filter(|&&c| intersects(&hints(&tree.nodes[c]), "comment reply"))
            .count()
            >= 2;
    }
    false
}
fn plausible(n: &Node, s: Stats, semantic: bool) -> bool {
    if s.characters == 0 {
        return semantic && s.images > 0;
    }
    if s.code > 0 || s.cells > 0 || n.tag == "p" && s.characters > s.linked {
        return true;
    }
    if s.density() >= 0.8 {
        return semantic && s.headings > 0 && s.blocks > 0;
    }
    has(CONTAINERS, &n.tag)
        || has(EVIDENCE, &n.tag)
        || has(HEADINGS, &n.tag)
        || has(STRUCTURES, &n.tag)
        || semantic
}
fn scores(tree: &Tree, stats: &[Stats], relaxed: bool) -> HashMap<Id, f64> {
    let mut result = HashMap::new();
    for id in tree.walk(0) {
        let n = &tree.nodes[id];
        let s = stats[id];
        if !has(CONTAINERS, &n.tag) || !plausible(n, s, false) {
            continue;
        }
        let tokens = hints(n);
        let mut score = (s.characters as f64).sqrt() + s.blocks.min(30) as f64 * 2.0;
        score += s.punctuation.min(40) as f64 * 0.25 + (s.code + s.cells).min(12) as f64 * 2.0;
        if intersects(&tokens, POSITIVE) {
            score += 8.0;
        }
        if !relaxed && intersects(&tokens, NEGATIVE) {
            score -= 18.0;
        }
        result.insert(id, score * (1.0 - s.density()).powi(2));
    }
    for id in tree.walk(0) {
        let n = &tree.nodes[id];
        if !has(EVIDENCE, &n.tag)
            || tree
                .elements(id)
                .iter()
                .any(|&c| has(EVIDENCE, &tree.nodes[c].tag))
        {
            continue;
        }
        let s = stats[id];
        let support = (1.0 + (s.characters as f64 / 100.0).min(3.0)) * (1.0 - s.density());
        let mut ancestor = n.parent;
        for weight in [1.0, 0.5, 0.25] {
            let Some(a) = ancestor else {
                break;
            };
            if let Some(score) = result.get_mut(&a) {
                *score += support * weight;
            }
            ancestor = tree.nodes[a].parent;
        }
    }
    result
}
fn score(ranked: &HashMap<Id, f64>, id: Id) -> f64 {
    ranked.get(&id).copied().unwrap_or(0.0)
}
fn winner(nodes: &[Id], ranked: &HashMap<Id, f64>) -> Option<Id> {
    let mut best = None;
    for &id in nodes {
        if best.is_none_or(|b| score(ranked, id) > score(ranked, b)) {
            best = Some(id);
        }
    }
    best
}
fn candidates(tree: &Tree, ranked: &HashMap<Id, f64>) -> Vec<Id> {
    tree.walk(0)
        .into_iter()
        .filter(|id| ranked.contains_key(id) && !has("body #document", &tree.nodes[*id].tag))
        .collect()
}
pub fn select(tree: &mut Tree) -> (Vec<Id>, Diagnostics) {
    visible(tree);
    let mut stats = statistics(tree);
    let mut removed = 0;
    let mut stack = vec![0];
    while let Some(id) = stack.pop() {
        let mut kept = Vec::new();
        for &c in &tree.nodes[id].children {
            if !tree.nodes[c].tag.is_empty() && clutter(tree, c, stats[c]) {
                removed += 1;
            } else {
                kept.push(c);
            }
        }
        stack.extend(
            kept.iter()
                .rev()
                .filter(|&&c| !tree.nodes[c].tag.is_empty())
                .copied(),
        );
        tree.nodes[id].children = kept;
    }
    if removed > 0 {
        stats = statistics(tree);
    }
    let mut ranked = scores(tree, &stats, false);
    let mut notes = Vec::new();
    if removed > 0 {
        notes.push(format!(
            "Removed {removed} conditionally identified clutter blocks."
        ));
    }
    let mut semantic = Vec::new();
    let mut landmarks = Vec::new();
    for id in tree.walk(0) {
        let n = &tree.nodes[id];
        let landmark = n.tag == "main" || n.attr("role") == "main";
        if !(landmark || n.tag == "article")
            || !plausible(n, stats[id], true)
            || clutter(tree, id, stats[id])
            || hints(n).contains("teaser") && stats[id].density() > 0.1
        {
            continue;
        }
        semantic.push(id);
        if landmark {
            landmarks.push(id);
        }
    }
    if !landmarks.is_empty() {
        semantic = landmarks;
    }
    if let Some(best) = winner(&semantic, &ranked) {
        return (
            vec![best],
            Diagnostics {
                strategy: Strategy::Semantic,
                notes,
            },
        );
    }
    let mut best = winner(&candidates(tree, &ranked), &ranked);
    if best.is_none_or(|b| score(&ranked, b) < 5.0) {
        notes.push("Retried selection once with relaxed class penalties.".into());
        ranked = scores(tree, &stats, true);
        best = winner(&candidates(tree, &ranked), &ranked);
    }
    if let Some(best) = best.filter(|&b| score(&ranked, b) >= 5.0) {
        let mut selected = vec![best];
        if let Some(parent) = tree.nodes[best].parent {
            let threshold = (score(&ranked, best) * 0.18).max(5.0);
            selected = tree
                .elements(parent)
                .into_iter()
                .filter(|&id| {
                    let n = &tree.nodes[id];
                    let s = stats[id];
                    !clutter(tree, id, s)
                        && (id == best
                            || score(&ranked, id) >= threshold && s.density() < 0.5
                            || has(HEADINGS, &n.tag) && s.density() < 0.5
                            || n.tag == "p" && s.density() < 0.25 && s.characters > 0)
                })
                .collect();
        }
        return (
            selected,
            Diagnostics {
                strategy: Strategy::Scored,
                notes,
            },
        );
    }
    let mut selected = Vec::new();
    let mut stack = vec![0];
    while selected.len() < 256 {
        let Some(id) = stack.pop() else {
            break;
        };
        let n = &tree.nodes[id];
        let s = stats[id];
        if clutter(tree, id, s) {
            continue;
        }
        let children = tree.elements(id);
        if (has(EVIDENCE, &n.tag) || has(HEADINGS, &n.tag) || has(STRUCTURES, &n.tag))
            && plausible(n, s, false)
            || children.is_empty() && s.characters > 0 && s.density() < 0.5
        {
            selected.push(id);
        } else {
            stack.extend(children.into_iter().rev());
        }
    }
    notes.push(
        if selected.is_empty() {
            "No relevant visible content found."
        } else {
            "Recovered plausible blocks."
        }
        .into(),
    );
    (
        selected,
        Diagnostics {
            strategy: Strategy::Fallback,
            notes,
        },
    )
}
